crates/accelerate/src/quantum_circuit/intern_context.rs

```diff
@@ -0,0 +1,144 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashMap;
+use pyo3::prelude::*;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::VecDeque;
+use std::hash::{Hash, Hasher};
+use std::mem::take;
+
+macro_rules! println {
+    ($($rest:tt)*) => {
+        #[cfg(debug_interner)]
+        std::println!($($rest)*)
+    }
+}
+
+#[cfg(debug_interner)]
+fn unique_id() -> u64 {
+    use std::sync::atomic::AtomicU64;
+    use std::sync::atomic::Ordering::SeqCst;
+    static COUNTER: AtomicU64 = AtomicU64::new(0);
+    COUNTER.fetch_add(1, SeqCst)
+}
+
+fn hash<T>(vec: &Vec<T>) -> u64
+where
+    T: Hash,
+{
+    let mut hasher = DefaultHasher::default();
+    vec.hash(&mut hasher);
+    hasher.finish()
+}
+
+pub type IndexType = u16;
+pub type BitType = u32;
```

Comment:
imo `u16` is too aggressive - we're not a million miles away from even all the _1q_ args not fitting in that, let alone the 2q+ ones.  On the surface, it doesn't make sense to me that `BitType` could be wider than `IndexType` - if we do that, we're saying that you can put more bits in a circuit than you can even apply 1q operations on.

I'm not certain that using `u16` actually saves any space anyway, because in the place it's used (`Vec<(PyObject, IndexType, IndexType)>`, the 8-byte alignment of `PyObject` means that we need to insert four bytes of padding into the tuple anyway.

Reply:
I'll try changing `IndexType` to `u32` and rerunning `memray`. I agree, even if `BitType` were also `u16`, it doesn't make sense for us to only store enough slots for all possible 1q arg lists. It may also make sense to have `CircuitData` skip using the interner when the total width of the arg list is smaller than the width of `BitType` (i.e. for 1q arg lists).

Reply:
I think the logic will get quite complex if we have multiple possible behaviours - I'd want to see that it's _definitely_ a problem before we went that route, I think.  Also in general, I don't think we care too much if small-scale stuff is a little bit bigger than it needs to be, it's the scaling to huge values that really matters.

Of course, we do construct quite a lot of small-scale circuits throughout Qiskit (gate-body definitions, etc), but the numbers of those aren't humongous, and all the work on singletons should drastically cut down their number as well.

Reply:
Done in eb35ffb. You were correct: no change in peak memory consumption when using the `u32` instead 😄.